File: 03 - source code/01 - imperative languages/Julia/random_streams_for_perf_stats.jl
#=
random_streams_for_perf_stats.jl

2026-01-01

run on Ubuntu 24 LTS: $ julia ./random_streams_for_perf_stats.jl

                      exe times measurements:
                      $ time julia -O1 ./random_streams_for_perf_stats.jl  => real	0m0.189s

                      $ multitime -n 20 julia -O1 ./random_streams_for_perf_stats.jl
                                  Mean        Std.Dev.    Min         Median      Max
                      real        0.191       0.002       0.186       0.191       0.195

                      $ multitime -n 20 julia ./random_streams_for_perf_stats.jl
                                  Mean        Std.Dev.    Min         Median      Max
                      real        0.202       0.002       0.196       0.202       0.206

                      $ multitime -n 20 julia -O0 ./random_streams_for_perf_stats.jl
                                  Mean        Std.Dev.    Min         Median      Max
                      real        0.168       0.004       0.164       0.167       0.179 <<<<<<<<<<<<<<<<<<<<<<


$ julia --version
julia version 1.12.3
$

=#

using Random

END = 62501  # 62501 for exactly 1M binary digits; const END has no speed effect here
# END = 11  # for testing

m = 65521  # = 2^16 - 15
a = 17364
c = 0

file_bits_x   = "random_bitstring.bin"
file_bits_hex = "random_bitstring.byte"


x = Array{UInt32}(undef, END)
x[1] = rand(1:m-1)  # indexing in Julia is 1-based; m is inclusive
# println("x[1] = ", x[1])  # for testing

# create large arrays of strings:
bits_x   = fill("000000000000", END-1)
bits_hex = fill("0000", END-1)


println("\ngenerating a random bit stream...")
for i in 2:END
    h = i - 1
    # println("h = $h")  # for testing
    x[i] = (a*x[h] + c) % m
    # println("x[i] = ", x[i])  # for testing

    bits_x_str0 = string(x[i], base=2)
    bits_x_str  = lpad(bits_x_str0, 16, "0")  # padding: https://www.jlhub.com/julia/manual/en/function/lpad
    # println("  bits_x_str =  ", bits_x_str)  # for testing
    bits_x[h]   = bits_x_str

    bits_hex_str0 = string(x[i], base=16)
    bits_hex_str  = lpad(bits_hex_str0, 4, "0")
    # println("  bits_hex_str =  ", bits_hex_str)  # for testing
    bits_hex[h]   = bits_hex_str
end


bits_x_str_total   = join(bits_x)
# println("\nbits_x_str_total = ", bits_x_str_total)  # for testing
bits_hex_str_total = join(bits_hex)
# println("\nbits_hex_str_total = ", bits_hex_str_total)  # for testing


# try to write the string to the file with error handling: Duck.ai
try
    # Open the file for writing
    open(file_bits_x, "w") do file
        write(file, bits_x_str_total)
    end
    println("Bit stream has been written to disk under name: ", file_bits_x)
catch e
    println("could not write to file: ", file_bits_x, " -- ", e)
end

try
    # Open the file for writing
    open(file_bits_hex, "w") do file
        write(file, bits_hex_str_total)
    end
    println("Byte stream has been written to disk under name: ", file_bits_hex)
catch e
    println("could not write to file: ", file_bits_hex, " -- ", e)
end


# end of random_streams_for_perf_stats.jl

File: 03 - source code/01 - imperative languages/Dart/random_streams_for_perf_stats.dart
/*
random_streams_for_perf_stats.dart

2026-02-02

run on Ubuntu 24 LTS: $ dart ./random_streams_for_perf_stats.dart
                      $ time dart ./random_streams_for_perf_stats.dart => real	0m0.151s
                      $ multitime -n 20 dart ./random_streams_for_perf_stats.dart
                      =>
                                    Mean        Std.Dev.
                        real        0.146       0.002    ==> so, this is the interpreted version, by the Dart VM

       JIT snapshot:  $ dart compile jit-snapshot random_streams_for_perf_stats.dart
                      $ dart run ./random_streams_for_perf_stats.jit
                      $ time dart run ./random_streams_for_perf_stats.jit => real	0m0.107s
                      
       AOT snapshot:  $ dart compile aot-snapshot ./random_streams_for_perf_stats.dart
                      $ dartaotruntime ./random_streams_for_perf_stats.aot
                      $ time dartaotruntime ./random_streams_for_perf_stats.aot => real	0m0.014s
                       

$ dart --version
Dart SDK version: 3.10.8 (stable) (Tue Jan 27 00:02:04 2026 -0800) on "linux_x64"
$

transpiled from random_streams_for_perf_stats.groovy with Duck.ai

*/


import 'dart:io';
import 'dart:math';

void main(List<String> args) {
  const int END = 62501;  // 62501 for exactly 1M binary digits

  const int m = 65521;  // = 2^16 - 15
  const int a = 17364;
  const int c = 0;

  String fileBitsX   = "random_bitstring.bin";
  String fileBitsHex = "random_bitstring.byte";

  List<int> x = List<int>.filled(END, 0);

  Random rnd = Random();
  x[0] = rnd.nextInt(m - 1) + 1;

  StringBuffer bitsX   = StringBuffer();
  StringBuffer bitsHex = StringBuffer();

  print("\ngenerating a random bit stream...");
  for (int i = 1; i < END; i++) {
    x[i] = (a * x[i - 1] + c) % m;

    String bitsXStr = x[i].toRadixString(2).padLeft(16, '0');
    bitsX.write(bitsXStr);

    String bitsHexStr = x[i].toRadixString(16).padLeft(4, '0');
    bitsHex.write(bitsHexStr);
  }


  // Write bit stream to disk:
  try {
    File(fileBitsX).writeAsStringSync(bitsX.toString());
    print("Bit stream has been written to disk under name:  $fileBitsX");
  } catch (e) {
    print("could not write to file: $fileBitsX ! -- ${e.toString()}");
  }

  // Write byte stream to disk:
  try {
    File(fileBitsHex).writeAsStringSync(bitsHex.toString());
    print("Byte stream has been written to disk under name: $fileBitsHex");
  } catch (e) {
    print("could not write to file: $fileBitsHex ! -- ${e.toString()}");
  }
}

// end of random_streams_for_perf_stats.dart
